src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/util/ObjectGraphUtils.java

```diff
@@ -190,24 +309,23 @@ private void pointersBecomeOneWayManyPairs(final Object[] fromPointers, final Ob
         for (int i = 0; i < fromPointers.length; i++) {
             fromToMap.put(fromPointers[i], toPointers[i]);
         }
-        final ObjectTracer pending = new ObjectTracer(false);
-        pointersBecomeOneWayFrames(pending, fromToMap);
-        final boolean currentMarkingFlag = pending.currentMarkingFlag;
-        AbstractSqueakObjectWithClassAndHash currentObject;
-        while ((currentObject = getNextFromWorkStack()) != null) {
-            if (currentObject.tryToMark(currentMarkingFlag)) {
-                currentObject.pointersBecomeOneWay(fromToMap);
-                pending.tracePointers(currentObject);
-            }
+
+        final ObjectTracer roots = ObjectTracer.fromRoots(image, false);
+        pointersBecomeOneWayFrames(roots, fromToMap);
+
+        final Runnable[] tasks = new Runnable[USABLE_THREAD_COUNT];
+        for (int i = 0; i < USABLE_THREAD_COUNT; i++) {
+            tasks[i] = new BecomeOneWayManyPairsTask(roots, fromToMap);
         }
+        runTasks(tasks);
     }
 
     @TruffleBoundary
     private static void pointersBecomeOneWayFrames(final ObjectTracer tracer, final Object fromPointer, final Object toPointer) {
-        Truffle.getRuntime().iterateFrames((frameInstance) -> {
+        final ContextObject resumeContextObject = Truffle.getRuntime().iterateFrames((frameInstance) -> {
             final Frame current = frameInstance.getFrame(FrameInstance.FrameAccess.READ_WRITE);
             if (!FrameAccess.isTruffleSqueakFrame(current)) {
-                return null;
+                return FrameAccess.getResumingContextObjectOrSkip(frameInstance);
```

Comment:
This is another issue I fixed along the way, here and in `pointersBecomeOneWayFrames(final ObjectTracer tracer, final UnmodifiableEconomicMap<Object, Object> fromToMap)`.